cmd/dep/ensure.go

```diff
@@ -116,9 +118,72 @@ func (cmd *ensureCommand) Run(ctx *dep.Ctx, args []string) error {
 	sm.UseDefaultSignalHandling()
 	defer sm.Release()
 
+	params := p.MakeParams()
+	if *verbose {
+		params.Trace = true
+		params.TraceLogger = log.New(os.Stderr, "", 0)
+	}
+	params.RootPackageTree, err = gps.ListPackages(p.AbsRoot, string(p.ImportRoot))
+	if err != nil {
+		return errors.Wrap(err, "ensure ListPackage for project")
+	}
+
+	if cmd.update {
+		applyUpdateArgs(args, &params)
+	} else {
+		err := applyEnsureArgs(args, cmd.overrides, p, sm, &params)
+		if err != nil {
+			return err
+		}
+	}
+
+	solver, err := gps.Prepare(params, sm)
+	if err != nil {
+		return errors.Wrap(err, "ensure Prepare")
+	}
+	solution, err := solver.Solve()
+	if err != nil {
+		handleAllTheFailuresOfTheWorld(err)
+		return errors.Wrap(err, "ensure Solve()")
+	}
+
+	sw := dep.SafeWriter{
+		Root:          p.AbsRoot,
+		Lock:          p.Lock,
+		NewLock:       solution,
+		SourceManager: sm,
+	}
+	if !cmd.update {
+		sw.Manifest = p.Manifest
+	}
+
+	// check if vendor exists, because if the locks are the same but
+	// vendor does not exist we should write vendor
+	vendorExists, err := dep.IsNonEmptyDir(filepath.Join(sw.Root, "vendor"))
+	if err != nil {
+		return errors.Wrap(err, "ensure vendor is a directory")
+	}
+	writeV := !vendorExists && solution != nil
+
+	return errors.Wrap(sw.WriteAllSafe(writeV), "grouped write of manifest, lock and vendor")
+}
+
+func applyUpdateArgs(args []string, params *gps.SolveParameters) {
+	// When -update is specified without args, allow every project to change versions, regardless of the lock file
+	if len(args) == 0 {
+		params.ChangeAll = true
+		return
+	}
+
+	// Allow any of specified project versions to change, regardless of the lock file
```

Comment:
I think we discussed checking for the presence of the named project in the lock file, and erroring out if it wasn't present. At least, that seems to make the most sense to me now, when looking at an implementation.